lib/ansible/modules/web_infrastructure/ansible_tower/tower_project.py

```diff
@@ -161,16 +161,19 @@ def main():
                     org = org_res.get(name=organization)
                 except (exc.NotFound) as excinfo:
                     module.fail_json(msg='Failed to update project, organization not found: {0}'.format(organization), changed=False)
-                try:
-                    cred_res = tower_cli.get_resource('credential')
-                    cred = cred_res.get(name=scm_credential)
-                except (exc.NotFound) as excinfo:
-                    module.fail_json(msg='Failed to update project, credential not found: {0}'.format(scm_credential), changed=False)
+                credential = None
+                if scm_credential:
+                    try:
+                        cred_res = tower_cli.get_resource('credential')
+                        cred = cred_res.get(name=scm_credential)
+                        credential = cred['id']
+                    except (exc.NotFound) as excinfo:
+                        module.fail_json(msg='Failed to update project, credential not found: {0}'.format(scm_credential), changed=False)
 
                 result = project.modify(name=name, description=description,
                                         organization=org['id'],
                                         scm_type=scm_type, scm_url=scm_url, local_path=local_path,
-                                        scm_branch=scm_branch, scm_clean=scm_clean, credential=cred['id'],
+                                        scm_branch=scm_branch, scm_clean=scm_clean, credential=credential,
```

Comment:
I think this follows the Ansible idempotency philosophy, but I may need help.

Say that the project _already_ had a credential associated with it. Task runs:

```yaml
- tower_project:
    name: some project
    organization: Default
    state: present
    scm_type: git
    scm_url: "{{ some_public_repository_url }}"
    tower_config_file: "{{ tower_cli_config_path }}"
```

Now, the existing credential will be removed.

I _think_ that's what's desired for modules. It's slightly counter-intuitive for me from an API perspective of the meaning of "blank".

If that's intended, then I will approve.

Reply:
Thank you for checking. I understand what you're saying, but I checked the behavior of the project, which already had a credential associated with it, on my machine and the existing credential was NOT removed.

```
TASK [Create a project] *********************************************************************************
ok: [awx-01]
```

Actually, the `scm_credential`'s behavior is same as the `scm_branch`'s one. I think there is no regression.